src/java.base/share/classes/jdk/internal/util/ClassFileDumper.java

```diff
@@ -0,0 +1,215 @@
+     * dumping of the generated classes, set the system property via
+     * -D<key>=<path>.
+     *
+     * The system property is read only once when it is the first time
+     * the dumper instance for the given key is created.
+     *
+     * If not enabled, this method returns ClassFileDumper with null
+     * dump path.
+     */
+    public static ClassFileDumper getInstance(String key) {
+        Objects.requireNonNull(key);
+
+        var dumper = DUMPER_MAP.get(key);
+        if (dumper == null) {
+            String path = GetPropertyAction.privilegedGetProperty(key);
+            Path dir;
+            if (path == null || path.trim().isEmpty()) {
+                dir = null;
+            } else {
+                dir = validateDumpDir(Path.of(path.trim()));
+            }
+            var newDumper = new ClassFileDumper(key, dir);
+            var v = DUMPER_MAP.putIfAbsent(key, newDumper);
+            dumper = v != null ? v : newDumper;
+        }
+        return dumper;
+    }
+
+    /**
+     * Returns a ClassFileDumper instance for the given key with a given
+     * dump path. To enable dumping of the generated classes
+     *     -D<key> or -D<key>=true
+     *
+     * The system property is read only once when it is the first time
+     * the dumper instance for the given key is created.
+     *
+     * If not enabled, this method returns ClassFileDumper with null
+     * dump path.
+     */
+    public static ClassFileDumper getInstance(String key, Path path) {
+        Objects.requireNonNull(key);
+        Objects.requireNonNull(path);
+
+        var dumper = DUMPER_MAP.get(key);
+        if (dumper == null) {
+            String value = GetPropertyAction.privilegedGetProperty(key);
+            boolean enabled = value != null && value.isEmpty()
+                                    ? true : Boolean.parseBoolean(value);
+            Path dir = enabled ? validateDumpDir(path) : null;
+            var newDumper = new ClassFileDumper(key, dir);
+            var v = DUMPER_MAP.putIfAbsent(key, newDumper);
+            dumper = v != null ? v : newDumper;
+        }
+
+        if (dumper.isEnabled() && !path.equals(dumper.dumpPath())) {
+            throw new IllegalArgumentException("mismatched dump path for " + key);
+        }
+        return dumper;
+    }
+
+    private final String key;
+    private final Path dumpDir;
+    private final AtomicInteger counter = new AtomicInteger();
+
+    private ClassFileDumper(String key, Path path) {
+        this.key = key;
+        this.dumpDir = path;
+    }
+
+    public String key() {
+        return key;
+    }
+    public boolean isEnabled() {
+        return dumpDir != null;
+    }
+
+    public Path dumpPath() {
+        return dumpDir;
+    }
+
+    public int incrementAndGetCounter() {
+        return counter.incrementAndGet();
+    }
+
+    public Path pathname(String internalName) {
+        return dumpDir.resolve(encodeForFilename(internalName) + ".class");
+    }
```

Comment:
Could be `private` I think

Reply:
It's public in case the caller who invokes `ClassDefiner::defineClass` wants to print the path to which the classfile is written.